APB Run Command

djzager opened the pull request:
This PR implements an `apb run` command that works even as a developer to run an APB in a specified namespace. There are some limitations:

- Since apb build, prepare, etc. all work on the current apb project...this does that too. So I have to be "in" the 'hello-world-apb' project in order to `apb-run` that APB.
- APBs that try to create anything above the permissions of a `$namespace:admin` will fail...hard.
- APBs that use `asb_encode_binding` will fail because we aren't passing the POD_NAME and POD_NAMESPACE to the pod we create...I can add this. But I think the right thing from an APB developer perspective is to specifically create a secret when not using the ansible broker. However, I'm not sure how we would communicate that to APBs.

djzager committed:
APB Run Command

djzager commented:
I still need to fix `apb-test`, cleanup the service accounts and role-bindings et al...but I wanted to get some eyeballs on this before I went through that trouble.

djzager committed:
Create a new apb run function

djzager committed:
Add POD_NAME and POD_NAMESPACE to created pods env

Even though APB developers should be using the `asb_encode_binding` only
when being run from the Broker, we should still give the APB this info.

dymurray commented on src/apb/cli.py:
```diff
@@ -387,6 +388,56 @@ def subcmd_test_parser(subcmd):
+    )
+
+    subcmd.add_argument(
+        '--registry',
```
Do we need this if we are already prompting for `tag`?

dymurray replied:
I noticed we have the same thing for build... Maybe it's okay. Just feels weird. I wouldn't be opposed to removing it from build either.

dymurray commented on src/apb/cli.py:
```diff
@@ -387,6 +388,56 @@ def subcmd_test_parser(subcmd):
+        help=u'Registry prefix of APB to prepend to tag'
+    )
+
+    subcmd.add_argument(
```
Same question as above.

dymurray commented on src/apb/engine.py:
```diff
@@ -464,84 +465,146 @@ def relist_service_broker(kwargs):
+        command=command,
+        service_account=sa
+    )
+
```
This might fail flake8 if you don't have 2 newlines.

djzager replied:
Fixed.

dymurray commented on src/apb/engine.py:
```diff
@@ -464,84 +465,146 @@ def relist_service_broker(kwargs):
         print("failed - %s" % e)
 
 
+def run_apb(project, image, name, action, parameters={}):
```
+1 I like how simple this function is.

dymurray commented on src/apb/engine.py:
```diff
@@ -707,6 +770,35 @@ def print_list(services):
+        exit(1)
+
+    name = "{}{}{}".format(
+        registry + "/" if registry is not None else "",
```
What happens if I set `--tag` and `--registry`?

djzager replied:
Looking at how `apb build` works. I took `tag` to be the name of the apb (since if it is not defined we take the name from the apb.yml). Not sure how we want to do this.

djzager committed:
Remove registry + org args in favor of tag

Add dockerfile argument to the `apb test` command (still not fixed).
Update command helpers to give full path example.

djzager committed:
Fix flake8 errors

djzager committed:
Wait for apb run to complete

Update apb run to watch the created pod and return final state.

djzager commented:
At this point, there are still some issues with `apb test` that I'm working through...but `apb run` should be done.

djzager committed:
Fix `apb test` command

Update the apb test to run in it's own namespace that we destroy after
the test.
Stop trying to wait for the test result if the apb commpletes
successfully or fails.
Add sleeps to our retrieve test results so we have some chance of the
test finishing.

djzager commented:
At this point `apb test` runs...but fails for some reason on the only apb for which we have a test playbook. This is worth investigation, but outside of the scope of this PR. 

https://github.com/ansibleplaybookbundle/mediawiki-apb/issues/7 Created

dymurray commented:
`apb run` worked for me first try! I'm really happy with this. Let's give folks a little more time to look at it but ACK from me.

djzager commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) Experiment with 'apb run' command for running APBs outside of Broker](https://trello.com/c/rZm87qDl/610-3-experiment-with-apb-run-command-for-running-apbs-outside-of-broker)

djzager commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) Experiment with 'apb run' command for running APBs outside of Broker](https://trello.com/c/rZm87qDl/610-3-experiment-with-apb-run-command-for-running-apbs-outside-of-broker)
